File: heap.py
class Heap:
    def __init__(self):
        self.heap = []
        self.size = 0

    def insert(self, value):
        self.heap.append(value)
        self.size += 1

        self.bubbleUpValue()

        print(self.heap)
    
    def bubbleUpValue(self):
        currentIndex = self.size - 1
        
        while currentIndex > 0 and self.heap[currentIndex] > self.heap[self.parentIndex(currentIndex)]:
            self.swap(currentIndex, self.parentIndex(currentIndex))
            currentIndex =  self.parentIndex(currentIndex)
            
    def swap(self, first, second):
        temp = self.heap[first]
        self.heap[first] = self.heap[second]
        self.heap[second] = temp

    def parentIndex(self, index):
        if index % 2 == 0:
            return (index - 2) / 2
        else:
            return (index - 1) / 2

    def remove(self):
        if self.size == 0:
            return False

        self.heap[0] = self.heap[self.size - 1]
        self.size -= 1

        self.bubbleDownValue()
        print(self.heap)

    def bubbleDownValue(self):
        currentIndex = 0

        while currentIndex <= self.size and not self.isValidParent(currentIndex):
            largerChildIndex = self.largerChildIndex(currentIndex)
            self.swap(currentIndex, largerChildIndex)
            currentIndex = largerChildIndex

    def isValidParent(self, index):
        if (not self.hasLeftChild(index)):
            return True

        if (not self.hasRightChild(index)):
            return (self.heap[index] >= self.leftChild(index))

        return (self.heap[index] >= self.leftChild(index) and 
        self.heap[index] >= self.rightChild(index))

    def largerChildIndex(self,index):
        if (not self.hasLeftChild(index)):
            return index

        if (not self.hasRightChild(index)):
            return self.leftChildIndex(index)

        if self.leftChild(index) > self.rightChild(index):
            return self.leftChildIndex(index)
        return self.rightChildIndex(index)

    def hasLeftChild(self, index):
        return self.leftChildIndex(index) <= self.size

    def hasRightChild(self, index):
        return self.rightChildIndex(index) <= self.size

    def leftChild(self, index):
        return self.heap[self.leftChildIndex(index)]

    def rightChild(self, index):
        return self.heap[self.rightChildIndex(index)]

    def leftChildIndex(self, index):
        return index * 2 + 1

    def rightChildIndex(self, index):
        return index * 2 + 2
            
    @classmethod
    def heapify(cls, givenArray):
        for index in range(len(givenArray)):
            cls.heapifyArray(givenArray, index)
        print(givenArray)
 
#   static methods can access static methods only
#   static methods call other static methods using className itself instead of self

    @classmethod
    def heapifyArray(cls, givenArray, index):
        largerIndex = index
        leftIndex = (index * 2) + 1
        rightIndex = (index * 2) + 2

        if leftIndex < len(givenArray) and givenArray[leftIndex] > givenArray[largerIndex]:
            largerIndex = leftIndex

        if rightIndex < len(givenArray) and givenArray[rightIndex] > givenArray[largerIndex]:
            largerIndex = rightIndex

        if largerIndex == index:
            return

        cls.swapHeapifyArray(givenArray, index, largerIndex)

        cls.heapifyArray(givenArray, largerIndex)

    @classmethod
    def swapHeapifyArray(cls, givenArray, index, largerIndex):
        temp = givenArray[index]
        givenArray[index] =  givenArray[largerIndex]
        givenArray[largerIndex] = temp

heap = Heap()
# heap.insert(10)
# heap.insert(5)
# heap.insert(17)
# heap.insert(4)
# heap.insert(22)

Heap.heapify([5,3,8,4,1,2])

File: heap.js
class Heap {
  constructor() {
    this.heap = [];
    this.size = 0;
  }

  insert = (value) => {
    this.heap[this.size] = value;
    this.size++;

    this.bubbleUpValue();

    console.log(this.heap);
  };

  bubbleUpValue = () => {
    var currentIndex = this.size - 1;

    while (
      currentIndex > 0 &&
      this.heap[currentIndex] > this.heap[this.parentIndex(currentIndex)]
    ) {
      this.swap(currentIndex, this.parentIndex(currentIndex));
      currentIndex = this.parentIndex(currentIndex);
    }
  };

  parentIndex = (index) => {
    if (index % 2 == 0) {
      return (index - 2) / 2; //formula of the parent node from a right Child in a heap
    } else {
      return (index - 1) / 2;
    } //formula of the parent node from a left Child in a heap
  };

  swap = (first, second) => {
    let temp = this.heap[first];
    this.heap[first] = this.heap[second];
    this.heap[second] = temp;
  };

  remove = () => {
    if (this.size == 0) return false;

    this.heap[0] = this.heap[this.size - 1];
    this.size--;

    this.bubbleDownValue();
    console.log(this.heap);
  };

  bubbleDownValue = () => {
    var currentIndex = 0;

    while (currentIndex <= this.size && !this.isValidParent(currentIndex)) {
      let largerChildIndex = this.largerChildIndex(currentIndex);
      this.swap(currentIndex, largerChildIndex);
      currentIndex = largerChildIndex;
    }
  };

  largerChildIndex = (index) => {
    if (!this.hasLeftChild(index)) return index;
    if (!this.hasRightChild(index)) return this.leftChildIndex(index);
    if (this.leftChild(index) > this.rightChild(index)) {
      return this.leftChildIndex(index);
    }
    return this.rightChildIndex(index);
  };

  isValidParent = (index) => {
    if (!this.hasLeftChild(index)) return true;
    if (!this.hasRightChild(index))
      return this.heap[index] >= this.leftChild(index);

    return (
      this.heap[index] >= this.leftChild(index) &&
      this.heap[index] >= this.rightChild(index)
    );
  };

  hasLeftChild = (index) => {
    return this.leftChildIndex(index) <= this.size;
  };

  hasRightChild = (index) => {
    return this.rightChildIndex(index) <= this.size;
  };

  leftChild = (index) => {
    return this.heap[this.leftChildIndex(index)];
  };

  rightChild = (index) => {
    return this.heap[this.rightChildIndex(index)];
  };

  leftChildIndex = (index) => {
    return index * 2 + 1;
  };

  rightChildIndex = (index) => {
    return index * 2 + 2;
  };

  static heapify = (givenArray) => {
    for (let index = 0; index < givenArray.length; index++) {
      Heap.heapifyArray(givenArray, index);
    }
    console.log(givenArray);
  };

  //static methods can access static methods only
  //static methods call other static methods using className itself instead of this

  static heapifyArray = (givenArray, index) => {
    let largerIndex = index;
    let leftIndex = index * 2 + 1;
    let rightIndex = index * 2 + 2;
    if (
      leftIndex < givenArray.length &&
      givenArray[leftIndex] > givenArray[largerIndex]
    ) {
      largerIndex = leftIndex;
    }
    if (
      rightIndex < givenArray.length &&
      givenArray[rightIndex] > givenArray[largerIndex]
    ) {
      largerIndex = rightIndex;
    }

    if (index == largerIndex) return;

    Heap.swapHeapifyElements(givenArray, index, largerIndex);

    Heap.heapifyArray(givenArray, largerIndex);
  };

  static swapHeapifyElements = (array, first, second) => {
    let temp = array[first];
    array[first] = array[second];
    array[second] = temp;
  };
}
let heapArray = new Heap();
// heapArray.insert(10);
// heapArray.insert(5);
// heapArray.insert(17);
// heapArray.insert(4);
// heapArray.insert(22);

Heap.heapify([5, 3, 8, 4, 1, 2]);
